pygsti/protocols/protocol.py

```diff
@@ -1683,7 +1778,11 @@ def __init__(self, circuits, qubit_labels=None):
         else:
             self.aux_info = {c: None for c in circuits}
         super().__init__(circuits, qubit_labels)
-        self.auxfile_types['aux_info'] = 'pickle'
+        
+        # Don't save all_circuits_needing_data, it's redundant with aux_info keys
+        self.auxfile_types['all_circuits_needing_data'] = 'reset'
+        # Currently not jsonable, but will be fixed in write()
```

Comment:
Would it be possible to add an assertion or other check here that check this is true? That is, that `self.aux_info.keys()` is equivalent to `all_circuits_needing_data`? This is definitely true initially upon construction, but it also looks like it is possible for these to become out of sync during the course of standard usage. For example, the base `ExperimentDesign` class has the method `truncate_to_circuits` which is a wrapper around `_truncate_to_circuits_inplace`. Looking at the implementation of `_truncate_to_circuits_inplace` this operates directly on `all_circuits_needing_data` and we never re-sync this with `aux_info`.  There are a couple other methods where this happens as well.

Reply:
Good point. I think probably other things would break if they got out of sync too, but an assert here to check is very easy.